include/pybind11/detail/internals.h

```diff
@@ -419,11 +419,33 @@ PYBIND11_NOINLINE internals &get_internals() {
     } gil;
     error_scope err_scope;
 
-    PYBIND11_STR_TYPE id(PYBIND11_INTERNALS_ID);
-    auto builtins = handle(PyEval_GetBuiltins());
-    if (builtins.contains(id) && isinstance<capsule>(builtins[id])) {
-        internals_pp = static_cast<internals **>(capsule(builtins[id]));
+    constexpr const char *id_cstr = PYBIND11_INTERNALS_ID;
+    str id(id_cstr);
 
+    dict state_dict;
+#if PY_VERSION_HEX < 0x03080000 || defined(PYPY_VERSION)
+    state_dict = reinterpret_borrow<dict>(PyEval_GetBuiltins());
+#elif PY_VERSION_HEX < 0x03090000
+    state_dict = reinterpret_borrow<dict>(PyInterpreterState_GetDict(_PyInterpreterState_Get()));
+#else
+    state_dict = reinterpret_borrow<dict>(PyInterpreterState_GetDict(PyInterpreterState_Get()));
+#endif
+
```

Comment:
I looked at the implementation of `PyInterpreterState_Get()` (below for easy reference), it has a path to set the Python error indicator. I think it's best to have `if` below here, to not clobber reporting of the root cause of the failure, but also because we have `assert(!PyErr_Occurred());` in `pybind11_fail()`:

```
if (PyErr_Occurred()) {
    throw error_already_set();
}
```

I'm not sure if we still need `if (!state_dict)` in the next line.

Maybe easiest and best:
```
if (PyErr_Occurred() || !state_dict) {
    throw error_already_set();
}
```

If there is an unexpected inconsistency, `error_already_set()` [will tell us clearly](https://github.com/pybind/pybind11/blob/0176632e8cef72a4223f2df54d6dfb9e552ec71f/include/pybind11/pytypes.h#L451-L454).

We could use [`raise_from(PyExc_SystemError, "get_internals(): could not acquire state dictionary!")`](https://github.com/pybind/pybind11/blob/0176632e8cef72a4223f2df54d6dfb9e552ec71f/include/pybind11/pytypes.h#L688)  (untested) but maybe that's overkill?

```
PyInterpreterState *
PyInterpreterState_Get(void)
{
    PyThreadState *tstate = _PyThreadState_GET();
    _Py_EnsureTstateNotNULL(tstate);
    PyInterpreterState *interp = tstate->interp;
    if (interp == NULL) {
        Py_FatalError("no current interpreter");
    }
    return interp;
}
```

Reply:
> We could use [raise_from(PyExc_SystemError, "get_internals(): could not acquire state dictionary!")](https://github.com/pybind/pybind11/blob/0176632e8cef72a4223f2df54d6dfb9e552ec71f/include/pybind11/pytypes.h#L688) (untested) but maybe that's overkill?

I like it, given that this fancy error reporting functionality already exists.